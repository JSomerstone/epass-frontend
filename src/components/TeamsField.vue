<template>
  <div>
    <b-field v-if="editable">
      <b-autocomplete
          v-model="name"
          v-if="editable"
          placeholder="Team name / Country"
          keep-first
          clear-on-select
          expanded
          :data="matchingNames"
          @keyup.native="onInput"
          @select="addTeam">
      </b-autocomplete>
      <b-input disabled expanded v-else />
      <b-button 
        icon-left="plus" 
        @click="addTeam(name)" 
        type="is-info"
        :disabled="name.length < 2"
      >
        Add
      </b-button>
    </b-field>
    <div class="team-list">
      <b-tag v-for="(team, index) in teams" v-bind:key="index"
          :closable="editable"
          aria-close-label="Remove team"
          size="is-medium"
          @close="removeTeam(index)"
        >
          {{ team }}
      </b-tag>
    </div>
  </div>
</template>
<style>
.team-list span {
  margin: 3px;
}
</style>
<script>
export default {
  data(){
    return {
      teams: this.value,
      name: "",
    };
  },
  props: {
    value: {
      type: Array,
      required: true,
    },
    editable: {
      type: Boolean,
      default: true,
    },
  },
  methods: {
    removeTeam(index) {
      this.teams.splice(index, 1);
      this.$emit('input', this.teams);
    },
    onInput(value) {
      value.code == "Enter" && this.addTeam();
    },
    addTeam(selected = "") {
      if (selected) {
        this.name = selected;
      }
      if (this.name.length <= 2 || this.isDuplicate) {
        return;
      }
      this.teams.push(this.name);
      this.teams.sort();
      this.$store.dispatch("tournaments/addTeam", { team: this.name } );
      this.name="";
      this.$emit('input', this.teams);
    },
  },
  computed: {
    teamPool: function() {
      return this.$store.getters['tournaments/teams'];
    },
    matchingNames: function() {
      return this.teamPool.filter( team => {
        return team
          .toString()
          .toLowerCase()
          .indexOf(this.name.toLowerCase()) >= 0
      });
    },
    isDuplicate: function () {
      return Boolean(this.teams.find( 
        team => team.toLowerCase() == this.name.toLowerCase())
      );
    },
  }
}
</script>