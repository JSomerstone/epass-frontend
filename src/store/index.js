import Vue from "vue";
import Vuex from "vuex";

import axios from "axios";
import config from "../config/index";

import tournaments from "./modules/tournaments";
import referees from "./modules/referees";
import countries from "./modules/countries";
import auth from "./modules/auth";

axios.defaults.baseURL = config.backendUrl;
Vue.use(Vuex);

const store = new Vuex.Store({
  modules: {
    tournaments,
    referees,
    countries,
    auth
  }
});

export default store;