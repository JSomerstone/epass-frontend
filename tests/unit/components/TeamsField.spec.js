import { mount, RouterLinkStub } from "@vue/test-utils";
import Vuex from "vuex";
import localVue from "../localVue";
import store from "../mockStore";

import TeamsField from "@/components/TeamsField";

describe("TeamsField", () => {
  let { tournaments } = store.modules;
  let value;

  beforeEach(() => {
    tournaments.state.teams = ["Team A", "Team B", "Team C", "Team D"];
    value = ["Team A", "Team B"];
  });

  it("Teams are removable when field is editable", async () => {
    const wrapper = mount(TeamsField, {
      localVue,
      store: new Vuex.Store(store),
      propsData: { value, editable: true },
      stubs: {
        RouterLink: RouterLinkStub
      }
    });

    expect(wrapper).toMatchSnapshot();
    let list = wrapper.find("div.team-list");
    expect(list.text()).toContain("Team A");
    expect(list.text()).toContain("Team B");
    expect(list.text()).not.toContain("Team C");

    await wrapper.find("a.delete").trigger("click");
    expect(list.text()).not.toContain("Team A");
  });

  it("New team can be added when field is editable", async () => {
    const vuexStore = new Vuex.Store(store);
    const wrapper = mount(TeamsField, {
      localVue,
      store: new Vuex.Store(store),
      propsData: { value, editable: true },
      stubs: {
        RouterLink: RouterLinkStub
      }
    });

    await wrapper.find('input').setValue("Team C");
    await wrapper.find('a.dropdown-item').trigger("click");
    expect(wrapper.find("div.team-list").text()).toContain("Team C");
  });

  it("Team can be added with Enter", async () => {
    const wrapper = mount(TeamsField, {
      localVue,
      store: new Vuex.Store(store),
      propsData: { value, editable: true },
      stubs: {
        RouterLink: RouterLinkStub
      }
    });

    const inputField = wrapper.find('input')
    await inputField.setValue("Team X");
    await wrapper.vm.onInput({ code: "Enter" });
    expect(wrapper.find("div.team-list").text()).toContain("Team X");
    expect(tournaments.actions.addTeam).toHaveBeenCalled();
  });
});