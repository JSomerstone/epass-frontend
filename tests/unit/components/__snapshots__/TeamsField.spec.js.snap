// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamsField Teams are removable when field is editable 1`] = `
<div>
  <div class="field has-addons">
    <!---->
    <div class="autocomplete control is-expanded">
      <div class="control is-clearfix"><input type="text" autocomplete="off" placeholder="Team name / Country" class="input">
        <!---->
        <!---->
        <!---->
      </div>
      <transition-stub name="fade">
        <div class="dropdown-menu" style="display: none;">
          <div class="dropdown-content" style="display: none;">
            <!----> <a class="dropdown-item"><span>
                            Team A
                        </span></a><a class="dropdown-item"><span>
                            Team B
                        </span></a><a class="dropdown-item"><span>
                            Team C
                        </span></a><a class="dropdown-item"><span>
                            Team D
                        </span></a>
            <!---->
            <!---->
          </div>
        </div>
      </transition-stub>
    </div> <button type="button" disabled="disabled" class="button is-info"><span class="icon is-small"><i class="mdi mdi-plus"></i></span> <span>
      Add
    </span>
      <!----></button>
    <!---->
  </div>
  <div class="team-list"><span class="tag is-medium"><span class="">
        Team A
    </span> <a role="button" aria-label="Remove team" tabindex="0" class="delete is-small"></a></span><span class="tag is-medium"><span class="">
        Team B
    </span> <a role="button" aria-label="Remove team" tabindex="0" class="delete is-small"></a></span></div>
</div>
`;
